File: neetcode/python/155.min-stack.py
#
# @lc app=leetcode id=155 lang=python3
#
# [155] Min Stack
#

# @lc code=start
class MinStack:

    def __init__(self):
        self.stack = []
        self.aux = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if not self.aux or val <= self.aux[len(self.aux) - 1]:
            self.aux.append(val)

    def pop(self) -> None:
        popped = self.stack.pop()
        if popped == self.aux[len(self.aux) - 1]:
            self.aux.pop()

    def top(self) -> int:
        n = len(self.stack)
        return self.stack[n - 1]        

    def getMin(self) -> int:
        return self.aux[len(self.aux) - 1]        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
# @lc code=end

File: neetcode/go/155.min-stack.go
/*
 * @lc app=leetcode id=155 lang=golang
 *
 * [155] Min Stack
 */

// @lc code=start
type MinStack struct {
	stack []int
	aux   []int
}

func Constructor() MinStack {
	minStack := MinStack{
		stack: []int{},
		aux:   []int{},
	}
	return minStack
}

func (this *MinStack) Push(val int) {
	this.stack = append(this.stack, val)
	nAux := len(this.aux)
	if nAux == 0 || val <= this.GetMin() {
		this.aux = append(this.aux, val)
	}
}

func (this *MinStack) Pop() {
	n := len(this.stack)
	nAux := len(this.aux)
	popped := this.stack[n-1]
	if popped == this.aux[nAux-1] {
		this.aux = this.aux[:nAux-1]
	}
	this.stack = this.stack[:n-1]
}

func (this *MinStack) Top() int {
	n := len(this.stack)
	return this.stack[n-1]
}

func (this *MinStack) GetMin() int {
	n := len(this.aux)
	return this.aux[n-1]
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(val);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */
// @lc code=end

